### ISSUE TITLE
Flow does not complain about wrong property accesses in ternary operator

### ISSUE THREAD

**Issue by philippotto** [facebook/flow#3854]
Flow won't complain about wrong property accesses within the boolean expression of the ternary operator.

```js
type Y = {
  k: number,
}

function f1(y: Y) {
  var x1 = y.j === 1;           // flow finds the error
  var x1 = y.j === 1 ? 0 : 1;   // flow does not complain
}
```

It's not the same issue as #3500, but these things might be related (both pertain to the ternary operator).

/edit: The same behaviour can be found in normal if-checks:

```
if (y.j === 1) { ... }
```

will also typecheck. I wonder if this is a "feature"? Of course, this can be a legitimate refinement, but when I operate with (exact) types, I want to be able to rely on Flow to warn me about accessing properties which don't exist in a given type.

**Comment by singpolyma-shopify** [comment]
Does this still happen if `Y = {| k: number |}` ?

**Comment by philippotto** [comment]
@singpolyma-shopify Yes, even with exact object types flow doesn't throw an error. For non-exact types, I could understand the point, that it shouldn't throw an error. However, for exact object types, it does not make any sense to not throw an error, in my opinion.

In both cases I would argue, that the user should use something more explicit like `if (y.hasOwnProperty("j"))` in order to check for not existing properties of a type.

At the moment, this behavior is really annoying, since I cannot trust flow completely when refactoring types. I still have to watch out for conditions like `if (y.j === "something") {...}` since flow doesn't complain about wrong type usages here.

**Comment by rsolomon** [comment]
I suspect this is related to #3887. I'm curious if anyone could comment on if this is expected behavior or not. It really makes some refactoring challenging and error-prone. For example, we recently migrated away from Immutable.js. In migrating from Immutable Lists to Arrays, we had a few instances of the following (as an example):

```
return list.size === 0 ? x : y;
```

`size` is not a valid property of an `Array`, but because it was referenced in a ternary, the unknown property references were not caught. If we were to put it in its own variable, it would have been caught:

```
const listSize = list.size;
return listSize === 0 ? x : y;
```

```
3: const listSize = list.size;
                         ^ property `size`. Property not found in
3: const listSize = list.size;
                    ^ Array
```

If this is an expected gotcha to look out for, it would be helpful if it were listed in the documentation. Something along the same lines as the "Array access is unsafe" section of the [Array type documentation](https://flow.org/en/docs/types/arrays/#toc-array-access-is-unsafe).

**Comment by cvoege** [comment]
Is there any progress on this? The fact that flow is just completely okay with you doing wrong property access makes it pretty useless for static analysis, especially in React where you have a lot of ternaryys.

**Comment by gkz** [comment]
We error in both examples now: https://flow.org/try/#1N4Igxg9gdgZglgcxALlAIwIZoKYBsD6uEEAztvhgE6UYCe+JADpdhgCYowa5kA0I2KAFcAtiRQAXSkOz9sADwxgJ+NPTbYuQ3BMnTZA+Y2yU4IwRO4A6SFBIrGVDGM7c+h46fNRLuKxJIGWh8MeT0ZfhYlCStpHzNsFBAMIQkIEQwJODAQfiEyfBE4eWw2fDgofDBMsAALfAA3KjgsXGxxZC4eAw0G-GhcWn9aY3wWZldu-g1mbGqJUoBaCRHEzrcDEgBrbAk62kXhXFxJ923d-cPRHEpTgyEoMDaqZdW7vKgoOfaSKgOKpqmDA+d4gB5fMA-P6LCCMLLQbiLOoYCqgh6-GDYRYIXYLSgkRZkCR4jpddwPfJLZjpOBkO4AX34kA0SRWxgABABNdkAXnZwAAOlB2eytsh2cIRDdeEL6UKhTAHso4NB2TAAIwAClo4s5AEp+UKRYD2fJ1bz2UMAFa8nl89UAbhFzpdIoA9G61UQAO5qipsEjsiS1bDskyUCCUI3sk3yABMFuttvt7IA-OyAAzs8WO50er0QX1sCDtCUQCTsyAiRi4FFQWW5EANEwkFVQJINDNWOMANgAzFYMyB6UA
